Add info about `!` and `impl Trait`

camelid opened the pull request:
Fixes #76094.

@rustbot modify labels: T-doc C-enhancement

camelid committed:
Add info about `!` and `impl Trait`

rust-highfive commented:
r? @sfackler

(rust_highfive has picked a reviewer for you, use r? to override)

camelid commented:
Note that this is still very work-in-progress, I just wanted to write something down to see if it's roughly correct.

camelid commented:
Cc @jyn514

jyn514 commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
 /// When writing your own traits, `!` should have an `impl` whenever there is an obvious `impl`
-/// which doesn't `panic!`. As it turns out, most traits can have an `impl` for `!`. Take [`Debug`]
+/// which doesn't `panic!`. The reason is that functions returning an `impl Trait` cannot have
+/// divergence, i.e., returning `!`, as their only possible code path. As an example, this code
```
This is only true _if `!` does not implement the trait_. Which is sort of the whole point of this documentation :P

jyn514 replied:
nit: 'cannot have divergence' -> 'cannot diverge'

jyn514 commented on library/std/src/primitive_docs.rs:
````diff
@@ -194,14 +194,47 @@ mod prim_bool {}
+/// ```
+///
+/// The reason is that, in the first example, there are many possible types for `!` to coerce
+/// to, because the function's return value is polymorphic. However, in the second example, the
````
I don't know what polymorphic means in this context. I thought that meant like a type constructor that can be instantiated with multiple different types.

camelid replied:
My understanding is that polymorphic just means "can be one of many concrete types".

camelid replied:
I could be wrong though.

jyn514 replied:
Hmm I'm not sure one way or another, but either way I think it would be more clear to say 'multiple possible return types'.

camelid replied:
How about:

```suggestion
/// to, because the function can return one of many concrete types. However, in the second example, the
```

jyn514 commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
+///
+/// The reason is that, in the first example, there are many possible types for `!` to coerce
+/// to, because the function's return value is polymorphic. However, in the second example, the
+/// other branch returns `0` which has a concrete type that `!` can be coerced to. See issue
```
Maybe mention that the concrete type is `i32` to avoid confusion around `{integer}`.

camelid replied:
Or in this case `u32` :P

jyn514 commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
+///
+/// [#36375]: https://github.com/rust-lang/rust/issues/36375
+///
+/// As it turns out, though, most traits can have an `impl` for `!`. Take [`Debug`]
```
Most -> all? Not sure if that's true but it seems like it should be.

camelid replied:
That's what I thought too; I would think you would be able to `impl` any trait for `!` since it can *never* :) exist. The body of it could just be `panic!("This shouldn't happen!")`.

jyn514 replied:
Or even simpler, `*self`, yeah.

camelid replied:
You mean because it can be coerced to any concrete type, so it can just return itself?

jyn514 replied:
`-> impl Trait` is not allowed in trait methods. So any trait it implemented would require returning some concrete type. Since ! can be coerced to any type, it would be coerced to the required type.

camelid committed:
cannot have divergence -> cannot diverge

jyn514 commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,9 +194,9 @@ mod prim_bool {}
-/// divergence, i.e., returning `!`, as their only possible code path. As an example, this code
-/// doesn't compile:
+/// which doesn't `panic!`. The reason is that functions returning an `impl Trait` cannot diverge,
+/// i.e., returning `!`, as their only possible code path. As an example, this code doesn't
```
returning -> return to be consistent with 'diverge'

camelid replied:
I think this is resolved.

camelid committed:
It's only an issue without an `impl Trait for !`

camelid commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
 /// #![feature(never_type)]
 /// # use std::fmt;
 /// # trait Debug {
-/// # fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result;
+/// #     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result;
 /// # }
```
Why is `Debug` defined here instead of imported?

camelid replied:
I'm going to delete this redefinition and import it instead.

camelid replied:
Looks like it's because of the orphan rule or something, so I had to revert my change :(

camelid committed:
Specify `0` of type `u32`

camelid committed:
Be more specific about polymorphic return types

I no longer say "polymorphic" since it's a bit ambiguous here.

camelid committed:
other branch -> `else` branch

camelid commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
+///
+/// The reason is that, in the first example, there are many possible types that `!` could coerce
+/// to, because the function can return one of many concrete types. However, in the second example,
+/// the `else` branch returns a `0` of type `u32`, which is a concrete type that `!` can be coerced
```
I feel like this might still be a little confusing. The `else` branch doesn't explicitly say `0u32`, it's just the compiler infers `u32` from the return type. Is there a better way to say this than "a `0` of type `u32`"?

jyn514 replied:
The way you said it now seems fine: "the compiler infers that the 0 has type u32".

camelid committed:
Import `Debug` instead of redefining it

camelid committed:
Explain why the `0` is a `u32`

camelid commented on library/std/src/primitive_docs.rs:
````diff
@@ -194,15 +194,46 @@ mod prim_bool {}
+/// ```
+///
+/// The reason is that, in the first example, there are many possible types that `!` could coerce
+/// to, because the function can return one of many concrete types. However, in the second example,
+/// the `else` branch returns a `0`, which the compiler infers from the return type to be of type
+/// `u32`, which is a concrete type that `!` can be coerced to. See issue [#36375] for more
````
This feels like a bit of a run-on sentence. I'm not sure how I could reformulate it though.

camelid committed:
Redefine `Debug` instead of importing it

This reverts commit 7e2548fe69ff5ec4e5e06c8c28351cbf2ebf7eee.

Now I know why it was redefined: it seems like it's potentially because
of the orphan rule. Here are the error messages:

error[E0119]: conflicting implementations of trait `std::fmt::Debug` for type `!`:
 --> src/primitive_docs.rs:236:1
  |
6 | impl Debug for ! {
  | ^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `core`:
          - impl std::fmt::Debug for !;

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/primitive_docs.rs:236:1
  |
6 | impl Debug for ! {
  | ^^^^^^^^^^^^^^^-
  | |              |
  | |              `!` is not defined in the current crate
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead

camelid committed:
Remove empty comment

jyn514 commented on library/std/src/primitive_docs.rs:
```diff
@@ -233,7 +233,10 @@ mod prim_bool {}
+/// # use std::fmt;
+/// # trait Debug {
+/// #     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result;
+/// # }
```
Can you post the error message you got in the commit message? Or possibly just pick a different trait to implement :P

camelid replied:
You mean put the error message in the commit message? Or post it here? I can do either, I just wasn't sure what you meant.

The error message is in the CI logs for 7e2548f by the way.

jyn514 replied:
Put the error in the commit message.
```
 error[E0119]: conflicting implementations of trait `std::fmt::Debug` for type `!`:
 --> src/primitive_docs.rs:236:1
  |
6 | impl Debug for ! {
  | ^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `core`:
          - impl std::fmt::Debug for !;

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/primitive_docs.rs:236:1
  |
6 | impl Debug for ! {
  | ^^^^^^^^^^^^^^^-
  | |              |
  | |              `!` is not defined in the current crate
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead
```
Hmm, seems like this will pop up for any trait from core you try to implement. The dummy `Debug` seems fine then.

camelid replied:
Okay, added the errors to the commit message (now e13a701).

jyn514 commented on library/std/src/primitive_docs.rs:
````diff
@@ -194,14 +193,48 @@ mod prim_bool {}
+/// ```
+///
+/// The reason is that, in the first example, there are many possible types that `!` could coerce
+/// to, because the function can return one of many concrete types. However, in the second example,
````
```suggestion
/// to, because many types implement `Add<u32>`. However, in the second example,
```

jyn514 commented on library/std/src/primitive_docs.rs:
```diff
@@ -194,14 +193,48 @@ mod prim_bool {}
+/// The reason is that, in the first example, there are many possible types that `!` could coerce
+/// to, because the function can return one of many concrete types. However, in the second example,
+/// the `else` branch returns a `0`, which the compiler infers from the return type to be of type
+/// `u32`, which is a concrete type that `!` can be coerced to. See issue [#36375] for more
```
```suggestion
/// `u32`. Since this is a concrete type, `!` gets coerced to `u32`. See issue [#36375] for more
```

camelid replied:
What about:

```suggestion
/// `u32`. Since `u32` is a concrete type, `!` can and will be coerced to it. See issue [#36375] for more
```

jyn514 replied:
Not a giant fan, but I don't think it's _less_ clear.

camelid replied:
"Since this is a concrete type" just sounded a bit confusing to me since it sounds kind of like we're referring to `!` and not `u32`.

camelid committed:
Improve wording

Co-authored-by: Joshua Nelson <joshua@yottadb.com>

camelid committed:
Improve `assert!` section in `bool` docs

jyn514 commented:
@bors r+ rollup

Thanks for the PR!

bors commented:
:pushpin: Commit 913354b8463a82497ebe586cfb9f5a4a218767ad has been approved by `jyn514`

<!-- @bors r=jyn514 913354b8463a82497ebe586cfb9f5a4a218767ad -->
<!-- homu: {"type":"Approved","sha":"913354b8463a82497ebe586cfb9f5a4a218767ad","approver":"jyn514"} -->

camelid committed:
Break line at 100 characters

camelid commented:
Oh well, it looks like the line break commit didn't get through in time ;)

jyn514 commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 55637f566993e2f8659aa09288bfc00b9965c524 has been approved by `jyn514`

<!-- @bors r=jyn514 55637f566993e2f8659aa09288bfc00b9965c524 -->
<!-- homu: {"type":"Approved","sha":"55637f566993e2f8659aa09288bfc00b9965c524","approver":"jyn514"} -->

jyn514 commented:
> Oh well, it looks like the line break commit didn't get through in time ;)

Bors will automatically abort the merge if you push new commits. It will _not_ however remove the S-waiting-on-bors tag :(

bors commented:
:hourglass: Testing commit 55637f566993e2f8659aa09288bfc00b9965c524 with merge c79747195683ac4842eaacf65e8e4cddac6f7e69...
<!-- homu: {"type":"BuildStarted","head_sha":"55637f566993e2f8659aa09288bfc00b9965c524","merge_sha":"c79747195683ac4842eaacf65e8e4cddac6f7e69"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36356)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36356","builder_name":"checks-azure"} -->

rust-log-analyzer commented:
Your PR [failed](https://github.com/rust-lang-ci/rust/runs/1055071521) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/c79747195683ac4842eaacf65e8e4cddac6f7e69/checks/1055071521/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/c79747195683ac4842eaacf65e8e4cddac6f7e69/checks/1055071521/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[group]Run exit 1
exit 1
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
##[endgroup]
##[error]Process completed with exit code 1.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

jyn514 commented:
```
curl: (56) SSLRead() return error -9806
clang+llvm-10.0.0-x86_64-apple-darwin/lib/clang/10.0.0/lib/darwin/libclang_rt.tsan_ios_dynamic.dylib: Lzma library error:  No progress is possible
tar: Error exit delayed from previous errors.

##[error]Bash exited with code '1'.
Finishing: Install clang
```
What a strange error ... Looks network related?

@bors retry
